### DISCUSSION TITLE
Thresholding by NaN returns a discolored plot

### DISCUSSION THREAD

**Discussion by john-lyne** [pyvista/pyvista#1758] [Q&A] [answered]
The code below yields the attached image. The issue I have is that I want the plot colored by the values of 'u', however they seem to be colored by NaN (hence the grey). This isn't a situation where I can use threshold_percent or threshold by specific values, as the mask values I'm using are from an external dataset.

It's likely I'm missing something obvious as it doesn't seem like a difficult problem. Any advice would be appreciated, thanks.

```py
import numpy as np
import pyvista as pv

#3D [nx,ny,nz]=[128,128,384] np array of velocity values
u = np.load('u_xyz.npy')

#3D np array of scalar values ranging from (0-4) representing different distinct regions of the u dataset
masks = np.load('masks10_10k3.npy')

#recti-linear z axis
geo = np.loadtxt('geo.info', dtype = np.float64)
z = geo[:,0]

#Domain info
lx, ly = 2*np.pi, np.pi
nx_total, ny_total  = 768,768

#Node data, sub-domain
nx, ny, nz = u.shape

#Full domain
x_full = np.linspace(0,lx,nx_total)
y_full = np.linspace(0,ly,ny_total )

#Select sub-domain
x = x_full[:nx]
y = y_full[:ny]

#Set indices where mask isn't 1 to NaN
indices = masks == 1
u[~indices] = np.nan

grid = pv.RectilinearGrid(x, y, z)

grid["u"] = u.flatten(order="F")

outline = grid.outline()
threshed = grid.threshold() #This should threshold out all NaN values
pv.set_plot_theme("document")
plotter = pv.Plotter()
plotter.add_mesh(outline)
plotter.add_mesh(threshed)
plotter.add_axes()
plotter.show_bounds(grid='front', location='outer', all_edges=True)
plotter.show()
```

![image](https://user-images.githubusercontent.com/77559890/139303215-23b5ce37-449e-4e6e-9307-8751a57948d2.png)

**Comment by banesullivan** [comment, marked as answer]
The problem here is that your data values lie on the nodes (points) of the mesh and the threshold filter operates cell-wise. What is happening is that cells along the boundary of your NaNs and your valid points are kept by the threshold filter. You can see this if you turn on edge visibility and zoom in.

Here is an example to demonstrate with similar data:

```py
import pyvista as pv
from pyvista import examples
import numpy as np

def generate_similar_data():
    # Use ctp to move from cell data to point data
    mesh = examples.load_channels().ctp()
    mesh['facies'] = np.array(mesh['facies'], dtype=float)
    idx = np.argwhere(mesh['facies'] == 0)
    mesh['facies'][idx] = np.nan
    return mesh

mesh = generate_similar_data()

# Remove NaN's
threshed = mesh.threshold()
threshed.plot(show_edges=True)
```



![Screen Shot 2021-11-19 at 10 15 12 AM](https://user-images.githubusercontent.com/22067021/142663850-a8b67b4d-1d19-4ee3-8457-1ce86f2cd05f.png)


Using the `all_scalars=True` option of the threshold filter mitigates this:

```py
mesh.threshold(mesh.get_data_range(), all_scalars=True).plot()
```
![Screen Shot 2021-11-19 at 10 18 47 AM](https://user-images.githubusercontent.com/22067021/142664332-c478b1f9-9049-43c2-a086-693631f26e5e.png)

**Comment by john-lyne** [comment]
Thanks for the reply, well written. This fixed my issue, although at first I would get the error:

```ValueError:` Setting `all_scalars=True` and designating `scalars` is incompatible.  Set one or the other but not both```

But updating pyvista seemed to fix this. Thanks.
